### ISSUE TITLE
logging output

### ISSUE THREAD

**Issue by ratijas** [bytesnake/telebot#4]
are there any plans to employ more sophisticated logging than `println!`?  for example it would be convenient to include logger field into the bot `struct` and a function to register a particular handler on user's choice like [slog](https://crates.io/crates/slog) or [env_logger](https://crates.io/crates/env_logger).

**Comment by bytesnake** [comment]
Yes, I've already tried to set some info! points throughout the library (although there aren't many places because the structure is very flat) but got this annoying error
```
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/home/lorenz/Dokumente/telebot/target/debug/examples/reply-0cd0654e2f4c3de8.0.o" "-o" "/home/lorenz/Dokumente/telebot/target/debug/examples/reply-0cd0654e2f4c3de8" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "/home/lorenz/Dokumente/telebot/target/debug/deps" "-L" "/usr/lib" "-L" "/usr/lib" "-L" "/usr/lib" "-L" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "/home/lorenz/Dokumente/telebot/target/debug/deps/libtelebot-cf49c3cc2f58f91b.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libserde_json-874010ab70f07926.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libdtoa-800e947d6adc2627.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libitoa-f4899e2d13429285.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libuuid-c9903c4e5440f806.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/librand-c9d9fbdab2355ee4.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libnum_traits-51f88ce8f32c0283.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/liberased_serde-170937c22d4f11de.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libserde-78170cde6f10a512.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libtokio_curl-e789b7dff0a5680d.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libcurl-407bb9e246fd1e31.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libopenssl_probe-1f67f65cf1fa0375.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libcurl_sys-77368722d64efd53.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libopenssl_sys-3723604732106f11.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/liblibz_sys-fd74ce539fd130c9.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libtokio_core-8af657de8192838d.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libscoped_tls-01b0e693814db5aa.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libmio-8f1a0753838947a5.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libnix-e29566907bcb6425.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libvoid-b9b5c2181d189492.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libbitflags-6c0c42f1f858d67c.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libslab-1381e1711ce95cec.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libnet2-99df4bd9f160d00b.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libcfg_if-5cf5b4dd25466da1.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/liblibc-325adfd3fcb4fe41.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/liblazycell-f0c2d7897740574b.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/libfutures-d5ed0f178e8ae39b.rlib" "/home/lorenz/Dokumente/telebot/target/debug/deps/liblog-2f7366be74992849.rlib" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-9a66b6a343d52844.rlib" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-9d79f761aa668a33.rlib" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-2beb731af7a6faec.rlib" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librand-6bc49e032a89c77d.rlib" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcollections-a2a467c3ca3b6479.rlib" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-ce7b9706e1719f27.rlib" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_jemalloc-4b74d6f2808677d3.rlib" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-95af4192ed69a1c8.rlib" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_unicode-e54225ff8f33e08f.rlib" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-cd0ca85e71f914ca.rlib" "/home/lorenz/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-0bf24067248742a8.rlib" "-l" "curl" "-l" "ssl" "-l" "crypto" "-l" "z" "-l" "util" "-l" "dl" "-l" "rt" "-l" "pthread" "-l" "gcc_s" "-l" "pthread" "-l" "c" "-l" "m" "-l" "rt" "-l" "util"
  = note: /home/lorenz/Dokumente/telebot/target/debug/examples/reply-0cd0654e2f4c3de8.0.o: In function `telebot::bot::{{impl}}::_fetch::{{closure}}':
          /home/lorenz/Dokumente/telebot/src/bot.rs:144: undefined reference to `telebot::bot::Bot::_fetch::_$u7b$$u7b$closure$u7d$$u7d$::__STATIC_FMTSTR::h647880600d153490'
          /home/lorenz/Dokumente/telebot/target/debug/examples/reply-0cd0654e2f4c3de8.0.o: In function `telebot::bot::{{impl}}::_fetch::{{closure}}':
          /home/lorenz/Dokumente/telebot/<log macros>:7: undefined reference to `telebot::bot::Bot::_fetch::_$u7b$$u7b$closure$u7d$$u7d$::_LOC::h5febc7b409984e0d'
```
Perhaps it'll help to look into the generated code but for now I haven't much time.

**Comment by bytesnake** [comment]
Here is a commit which triggers the compile error: 28aaf82339a3b8183231160a0a12f5944eb135ab

**Comment by Eroc33** [comment]
@bytesnake in regards to your error for some reason `format!` in `impl trait` types doesn't work at the moment (see https://github.com/Nemo157/failing-impl-trait-thing ).

**Comment by bytesnake** [comment]
Hey, I've just retried to compile the the Debugging branch and the impl trait bug seems to be gone. Therefore I've merged the debugging branch #9. If you need additional logging points, please create a pull request. For now I'll close this issue.
